### PR TITLE
Client library cleanups

### PR DESCRIPTION
**Pull request by ash2k**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Please see individual commits.

/sig api-machinery
/release-note-none

### CODE REVIEW

**Comment by fedebongio** [comment]
/assign @sttts 
would you mind taking a look? not urgent! thanks

File: staging/src/k8s.io/client-go/tools/clientcmd/client_config.go [hunk 1]
```diff
@@ -193,13 +193,13 @@ func (config *DirectClientConfig) ClientConfig() (*restclient.Config, error) {
 		if err != nil {
 			return nil, err
 		}
-		mergo.MergeWithOverwrite(clientConfig, userAuthPartialConfig)
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:196]: MergeWithOverwrite is deprecated?

Comment (ash2k) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:196]: Yes https://github.com/imdario/mergo/blob/master/merge.go#L296

File: staging/src/k8s.io/client-go/tools/clientcmd/client_config.go [hunk 2]
```diff
@@ -542,11 +542,12 @@ func (config *inClusterClientConfig) RawConfig() (clientcmdapi.Config, error) {
 }
 
 func (config *inClusterClientConfig) ClientConfig() (*restclient.Config, error) {
-	if config.inClusterConfigProvider == nil {
-		config.inClusterConfigProvider = restclient.InClusterConfig
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:546]: this removes a side-effect, right? So formally it is a breaking change, although to the better.

@liggitt wdyt?

File: staging/src/k8s.io/client-go/tools/clientcmd/loader.go [hunk 3]
```diff
@@ -283,20 +281,14 @@ func (rules *ClientConfigLoadingRules) Migrate() error {
...
 		}
 
-		in, err := os.Open(source)
+		data, err := ioutil.ReadFile(source)
 		if err != nil {
 			return err
 		}
-		defer in.Close()
-		out, err := os.Create(destination)
+		err = ioutil.WriteFile(destination, data, 0666)
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/clientcmd/loader.go:288]: why these permissions?

Comment (ash2k) [reply staging/src/k8s.io/client-go/tools/clientcmd/loader.go:288]: This is what `os.Create()` uses under the hood https://github.com/golang/go/blob/go1.15.5/src/os/file.go#L316

Comment (sttts) [reply staging/src/k8s.io/client-go/tools/clientcmd/loader.go:288]: > it is created with mode 0666 (**before umask**)

Can you add a comment that this is umask'ed? I know our security ppl. finding this and complain.

Comment (ash2k) [reply staging/src/k8s.io/client-go/tools/clientcmd/loader.go:289]: Done + rebased.

File: staging/src/k8s.io/client-go/tools/clientcmd/client_config.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:201]: we seem to be ping-ponging back and forth on this library, which is unfortunate - https://github.com/kubernetes/kubernetes/pull/65779

File: staging/src/k8s.io/client-go/tools/clientcmd/client_config.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:552]: this looks fine. the type is private, the member field is only used from this method (outside of unit tests), and the default impl (`restclient.InClusterConfig`) is stateless

File: staging/src/k8s.io/client-go/tools/clientcmd/client_config.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:201]: @ash2k could you track why we didn't go your route in #65779 ? As far as I see the override option already existed back then, added in 2017 or so.

Comment (ash2k) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:201]: @sttts You mean why a deprecated function was introduced? I don't know and it wasn't discussed. The change in this PR does not change what is happening under the hood, as the current codepath just delegates to the new codepath - see https://github.com/imdario/mergo/blob/v0.3.5/merge.go#L194-L219

Comment (sttts) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:201]: no, why did we switch to a deprecated function in the PR @liggitt referenced. If it has been just an oversight, his argument of ping-pong goes away.

Comment (ash2k) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:201]: @sttts In #65779 we switched from `Merge()` to `MergeOverride()` for some reason. It was a change in behavior. Here I'm not changing the current behavior, I'm just replacing `MergeOverride()` with `Merge(WithOverride())` which are 100% identical (see the link above).

I think it was just an oversight that a deprecated function was used i.e. there is no ping-pong :)

Comment (sttts) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:201]: :+1:

Comment (sttts) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:201]: So I agree we should move away from the deprecated func and move forward.

**Pull request merged**